src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala

```diff
@@ -0,0 +1,151 @@
+    case SubscribeToWorkflow(id) =>
+      //  NOTE: This fails silently. Currently we're ok w/ this, but you might not be in the future
+      workflowStore.toMap.get(id) foreach {_ ! SubscribeTransitionCallBack(sender())}
+  }
+
+  private def workflowOutputs(id: WorkflowId): Future[binding.WorkflowOutputs] = {
+    workflowStore.toMap.get(id) map workflowToOutputs getOrElse Future.failed(new WorkflowNotFoundException)
+  }
+
+  private def workflowToOutputs(workflow: WorkflowActorRef): Future[binding.WorkflowOutputs] = {
+    workflow.ask(GetOutputs).mapTo[binding.WorkflowOutputs]
+  }
+
+  private def submitWorkflow(wdlSource: WdlSource, wdlJson: WdlJson, inputs: binding.WorkflowRawInputs,
+                             isRestart: Boolean, maybeWorkflowId: Option[WorkflowId]): Future[WorkflowId] = {
+    for {
+      namespace <- Future(WdlNamespace.load(wdlSource))
+      coercedInputs <- Future.fromTry(namespace.coerceRawInputs(inputs))
+      descriptor = new WorkflowDescriptor(namespace, wdlSource, wdlJson, coercedInputs)
+      workflowActor = context.actorOf(WorkflowActor.props(descriptor, backend, dataAccess))
+      workflowId = maybeWorkflowId.getOrElse(UUID.randomUUID())
+      _ <- Future.fromTry(workflowStore.insert(workflowId, workflowActor))
+    } yield {
+      workflowActor ! (if (isRestart) Restart else Start)
+      workflowActor ! SubscribeTransitionCallBack(self)
+      workflowId
+    }
+  }
+
+  private def restartWorkflow(restartableWorkflow: RestartableWorkflow): Future[WorkflowId] = {
+    submitWorkflow(restartableWorkflow.source, restartableWorkflow.json, restartableWorkflow.inputs,
+      isRestart = true, Option(restartableWorkflow.id))
+  }
+
+  private def updateWorkflowState(workflow: WorkflowActorRef, state: WorkflowState): Future[Unit] = {
+    val id = idByWorkflow(workflow)
+    dataAccess.updateWorkflowState(id, state)
+  }
+
+  private def idByWorkflow(workflow: WorkflowActorRef): WorkflowId = {
+    workflowStore.toMap.collectFirst { case (k, v) if v == workflow => k }.get
+  }
+
+  private def restartIncompleteWorkflows(): Unit = {
+    // If the clob inputs for this workflow can be converted to JSON, return the JSON
+    // version of those inputs in a Some().  Otherwise return None.
+    def clobToJsonInputs(workflowInfo: WorkflowInfo): Option[binding.WorkflowRawInputs] = {
+      workflowInfo.wdlJson.parseJson match {
+        case JsObject(rawInputs) => Option(rawInputs)
+        case x =>
+          log.error(s"Error restarting workflow ${workflowInfo.workflowId}: expected JSON inputs, got '$x'.")
+          None
+      }
+    }
+
+    type QuantifierAndPlural = (String, String)
+    def pluralize(num: Int): QuantifierAndPlural = {
+      (if (num == 0) "no" else num.toString, if (num == 1) "" else "s")
+    }
+
+    val result = for {
+      workflowInfos <- dataAccess.getWorkflowsByState(Seq(WorkflowSubmitted, WorkflowRunning))
```

Comment:
Similar comment on the single clause for comprehension. Also the nested for comprehensions makes me a little twitchy.